lonboard/_geoarrow/movingpandas_interop.py

```diff
@@ -0,0 +1,199 @@
+
+
+# TODO (lonboard-specific):
+# - update timestamp serialization to cast to float32 at that point
+# # offset by earliest timestamp
+# timestamps -= timestamps.min()
+
+# # Cast to float32
+# timestamps = timestamps.astype(np.float32)
+
+
+def movingpandas_to_geoarrow(
+    traj_collection: TrajectoryCollection,
+) -> Tuple[Table, ChunkedArray]:
+    """Convert a MovingPandas TrajectoryCollection to GeoArrow
+
+    Args:
+        traj_collection: _description_
+
+    Returns:
+        _description_
+    """
+    import pyarrow as pa
+    import shapely
+
+    crs = traj_collection.get_crs()
+    crs_json = crs.to_json_dict() if crs is not None else None
+
+    num_coords = 0
+    num_trajectories = len(traj_collection)
+    offsets = np.zeros(num_trajectories + 1, dtype=np.int32)
+    datetime_dtypes = set()
+    attr_schemas: List[pa.Schema] = []
+
+    # Loop the first time to infer offsets for each trajectory
+    for i, traj in enumerate(traj_collection.trajectories):
+        traj: mpd.Trajectory
+
+        num_coords += traj.size()
+        offsets[i + 1] = num_coords
+        datetime_dtypes.add(traj.df.index.dtype)
+
+        geom_col_name = traj.get_geom_col()
+        df_attr = traj.df.drop(columns=[geom_col_name])
+
+        # Explicitly drop index because the index is a DatetimeIndex that we convert
+        # manually later.
+        arrow_schema = pa.Schema.from_pandas(df_attr, preserve_index=False)
+        attr_schemas.append(arrow_schema)
+
+    assert (
+        len(datetime_dtypes) == 1
+    ), "Expected only one datetime dtype across all trajectories."
+    datetime_dtype = list(datetime_dtypes)[0]
+
+    # We currently always provision space for XYZ coordinates, and then only use 2d if
+    # the Z dimension is always NaN
+    coords = np.zeros((num_coords, 3), dtype=np.float64)
+
+    # Infer an arrow time unit from the numpy
+    time_unit, time_arrow_dtype = infer_time_unit(datetime_dtype)
+
+    # TODO: switch this to just using `time_arrow_dtype.bit_width` once
+    # https://github.com/kylebarron/arro3/pull/190 is released
+    if time_unit in {"s", "ms"}:
+        timestamps = np.zeros(num_coords, dtype=np.int32)
+    elif time_unit in {"us", "ns"}:
+        timestamps = np.zeros(num_coords, dtype=np.int64)
+    else:
+        raise ValueError(f"Unexpected time unit: {time_unit}.")
+
+    attr_schema = pa.unify_schemas(attr_schemas, promote_options="permissive")
+    attr_tables: List[pa.Table] = []
+
+    # Loop second time to fill timestamps and coords
+    for i, traj in enumerate(traj_collection.trajectories):
+        start_offset = offsets[i]
+        end_offset = offsets[i + 1]
+
+        timestamps[start_offset:end_offset] = traj.df.index
+        coords[start_offset:end_offset, 0] = shapely.get_x(traj.df.geometry)
+        coords[start_offset:end_offset, 1] = shapely.get_y(traj.df.geometry)
+        coords[start_offset:end_offset, 2] = shapely.get_z(traj.df.geometry)
```

Comment:
```suggestion
        coords[start_offset:end_offset] = shapely.get_coordinates(traj.df.geometry, include_z=True)
```

This might work as well? (and be more efficient)

Reply:
> Interestingly... `shapely.get_coordinates` is 3x faster than `get_x` in isolation!

I was going to say that this is expected, because `get_coordinates` can do cheaper memcpy of the full coordinate sequence buffer. But, I would actually only expect to see a significant difference there for geometries with longer CoordSeqs, so not for points ..

Now, if I test this with a random array, I do see that an individual `get_x` is faster:

```
In [10]: arr = shapely.points(np.random.randn(10000, 2))

In [11]: %timeit shapely.get_x(arr)
256 µs ± 25.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [13]: %timeit shapely.get_coordinates(arr)
966 µs ± 117 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

So in practice I think what you are seeing above is mostly a slightly different overhead between both options for small arrays. In the end, it's probably only a code simplication and so not much of a performance improvement

Reply:
> TIL that there's a considerable difference between applying shapely operations on a `GeometryArray` vs a `GeoSeries`:

Yeah, that is because in the GeoSeries case it reconstructs a (Geo)Series from the result (at least for ufuncs like `get_x`, not for `get_coordinates`), which has some overhead. And in this case you don't need a series, but just an array and so can avoid that overhead by calling the shapely function on the array